Enable Floating point exceptions by default in Debug mode (GCC/Clang)

jmarrec opened the pull request:
Pull request overview
---------------------

Realize there was still a TODO... Let's see if CI complains or not now.

* Fixes #10648 


### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

jmarrec committed:
Enable Floating point exceptions by default in Debug mode (?)

Myoldmopar commented:
Oh yes, CI did complain :)  https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-91879baaf9bdf714a58a4df7e4b60993d6e70d6d-x86_64-Linux-Ubuntu-22.04-gcc-11.4-UnitTestsCoverage-Debug.html

jmarrec commented:
@Myoldmopar Shouldn't these errors be investigated and fixed (either the code needs fixing or the test initialization itself)?

Myoldmopar commented:
> @Myoldmopar Shouldn't these errors be investigated and fixed (either the code needs fixing or the test initialization itself)?

Yes, are you asking me to do this investigation and fix things?  If so, I can add it to my list and get to it soon.  If not, are you wanting to?  

And as an aside, we are in the unfortunate position of including a third party library that uses NaN intentionally as a signal value.  I'm not sure what will come of this, but I wanted to offer a small heads up here.

jmarrec committed:
Fix RefrigeratedRackWithCaseInZone by initializing OutBaroPress

jmarrec committed:
CurveManager: when divisor is zero, don't try to exercise it

Fixes  EnergyPlusFixture.DivisorNormalizationDivisorOnlyButItIsZero

jmarrec committed:
Fix CO2ControlDesignOccupancyTest

The issue was first unitialized StdRhoAir and co, but the main problem was getting the test to pass. Took me a while to figure out...

jmarrec committed:
Fix WindowManager_RefAirTempTest by properly allocating arrays

jmarrec committed:
Work around for DXCoils_Test1

cf #10651

jmarrec committed:
Fix HPWHSizing

jmarrec committed:
Fix VariableSpeedCoils_ZeroRatedCoolingCapacity_Test

cf https://github.com/NREL/EnergyPlus/pull/10470#pullrequestreview-2230090029
@RKStrand does that mean sense to you please?

jmarrec committed:
Fix SolarShadingTest_CalcBeamSolarOnWinRevealSurface

jmarrec committed:
Fix ManageElectricPowerTest_CheckOutputReporting

jmarrec committed:
Fix UnitaryBypassVAV_ParentElectricityRateTest

jmarrec committed:
Fix HVACMSHP_UnitarySystemElectricityRateTest

jmarrec committed:
Similar fix for PTHP test: init the RatedAirMassFlowRate due to sizing disabled

jmarrec committed:
Fix TestReadingCoilCoolingHeatingDX

jmarrec committed:
Fix SingleSpeedDXCoolingCoilOutputTest

jmarrec committed:
Fix MultiSpeedDXCoolingCoilOutputTest

jmarrec committed:
Fix SQLiteFixture.DXCoils_TestComponentSizingOutput_SingleSpeed / DXCoils_TestComponentSizingOutput_TwoSpeed

jmarrec committed:
Fix  EnergyPlusFixture.TwoSpeedDXCoilStandardRatings_Curve_Fix_Test EnergyPlusFixture.TwoSpeedDXCoilStandardRatingsTest

Myoldmopar commented:
After @jmarrec enabled floating point exception catching on Mac, and found a whole bunch of failing tests, he got to work fixing them up.  There are unfortunately 5 that are still remaining, and because of their nature (sizing, coils, ...), it would be great to get insight from @rraustad .  I'm inclined to just disable the FPE on this branch long enough to merge it in, then open a new branch where they fail again, and resolve them there.

Thoughts anyone?

rraustad commented:
So what is the method to find these FP exceptions? Run unit tests in a debug build? That's very painful.

rraustad commented:
This unit test runs locally in a Windows debug build (both with and without debugger):

EnergyPlusFixture.DXCoils_Test2 (DXCoils_Test1 shown in #10648 also passes)

    Note: Google Test filter = EnergyPlusFixture.DXCoils_Test2
    [==========] Running 1 test from 1 test suite.
    [----------] Global test environment set-up.
    [----------] 1 test from EnergyPlusFixture
    [ RUN ] EnergyPlusFixture.DXCoils_Test2

![image](https://github.com/user-attachments/assets/36822b2f-c7ef-49e8-bc44-8bd0fd251b2f)

jmarrec committed:
Catch floating point exceptions on MSVC for gtest + add cmake `FORCE_DEBUG_ARITHM_MSVC` to enable in release builds too

jmarrec commented:
I can't believe how long this took me, but I think I got it working for windows too, in https://github.com/NREL/EnergyPlus/commit/64018cfae450593e5e0f4077ec49f2d5bd9a47ef

Turns out the gtest was not enabling floating point exceptions for msvc (ever).

I added a `FORCE_DEBUG_ARITHM_MSVC` cmake configuration variable that should allow turning it on for MSVC release builds too, and by default it'll enable it in Debug builds.

I'm testing it on MSVC now, will consider adding it to this branch if it behaves properly.

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into enable_fe_default_debug

jmarrec commented:
Alright, confirmed it appears to work fine on MSVC, so I added it to this branch.

```
$ devenv /debugexe .\Products\energyplus_tests.exe '--gtest_filter=*DXCoils_Test2*'
```

![image](https://github.com/user-attachments/assets/545ce7e6-f603-426b-8673-ed9261eb5323)


In release, after calling `cmake -DFORCE_DEBUG_ARITHM_MSVC:BOOL=ON . `

```
$ ctest -R DXCoils_Test2 --output-on-failure
Test project C:/src/EnergyPlus/build-release
    Start 1095: EnergyPlusFixture.DXCoils_Test2
1/1 Test #1095: EnergyPlusFixture.DXCoils_Test2 ...***Exception: Numerical  0.31 sec
Note: Google Test filter = EnergyPlusFixture.DXCoils_Test2
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from EnergyPlusFixture
[ RUN      ] EnergyPlusFixture.DXCoils_Test2


0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.56 sec

The following tests FAILED:
        1095 - EnergyPlusFixture.DXCoils_Test2 (NUMERICAL)
Errors while running CTest
```

Myoldmopar commented:
@jmarrec this is great, thanks for another major effort.  

@rraustad could you confirm that these tests fail locally for you now?  I'd love to know whether to just wait here until they are finally fixed and merge this all at once, or get this merged in as nearly-done and finish on a separate branch.

Thanks all!

rraustad commented:
@Myoldmopar working as expected. I could push the fix for DXCoils_Test2 which also fixes TestMultiSpeedCoilsAutoSizingOutput. The MultispeedDX* failures may take a little time to figure out the Nan.

![image](https://github.com/user-attachments/assets/3b41880d-555e-4f01-8177-349f9daaf0fa)

jmarrec commented:
@rraustad feel free to push indeed! thanks

rraustad committed:
Fix 5 DX coil unit tests

Myoldmopar commented:
Down to 9!  Looks like they are 3 polygon and 6 ITE, hopefully just two more one-liners! :)

rraustad commented:
ITE* unit test, old favorite:

![image](https://github.com/user-attachments/assets/271a4f9e-f0da-4e23-8f7e-6f98a49d9d64)

rraustad committed:
Fixed crash, now get diffs in solar shading

rraustad commented:
I have no idea why fixing the crash causes the SurfSunlitFrac values to change. 1) the non-zero values are for the mirrored shading surface now, and 2) the south was has a 0 sunlit fraction ??

rraustad commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -3947,11 +3947,11 @@ void CLIPLINE(Real64 &x1, Real64 &x2, Real64 &y1, Real64 &y2, Real64 maxX, Real6
                     visible = true;
                     x1 = minX;
-                    y1 = maxY + e / dx;
+                    y1 = maxY + General::SafeDivide(e, dx);
```
There may be a better way to correct this.

jmarrec replied:
* This is fixed separately in #10657

rraustad commented on tst/EnergyPlus/unit/AirflowNetworkHVAC.unit.cc:
```diff
@@ -2212,7 +2212,7 @@ TEST_F(EnergyPlusFixture, AirflowNetwork_TestPressureStat)
     // Check indoor pressure and mass flow rate
     EXPECT_NEAR(PressureSet, state->afn->AirflowNetworkNodeSimu(3).PZ, 0.0001);
-    EXPECT_NEAR(0.00255337, state->afn->ReliefMassFlowRate, 0.0001);
+    EXPECT_NEAR(0.06551, state->afn->ReliefMassFlowRate, 0.0001);
```
I assume since StdRhoAir is now in the fixture that something has changed BEFORE this unit test sets StdRhoAir.

rraustad commented:
Looks like I forgot to add the FPE fix for PolygonClippingDirect

    state->dataBSDFWindow->SUNCOSTS(4, 9)(1) = 0.2; <-- will be Nan in FigureSolarBeanAtTimestep if not initialized
    state->dataBSDFWindow->SUNCOSTS(4, 9)(2) = 0.2;
    state->dataBSDFWindow->SUNCOSTS(4, 9)(3) = 0.2;
    FigureSolarBeamAtTimestep(*state, state->dataGlobal->HourOfDay, state->dataGlobal->TimeStep);

and looking at the reason for the CI failures again for SolarShadingTest_PolygonOverlap2 and 3.

The south wall has coordinates:

        "    -5,-5,3,  !- X,Y,Z ==> Vertex 1 {m}",
        "    -5,-5,0,  !- X,Y,Z ==> Vertex 2 {m}",
        "    5,-5,0,  !- X,Y,Z ==> Vertex 3 {m}",
        "    5,-5,3;  !- X,Y,Z ==> Vertex 4 {m}",

and the south shade nearest the wall has coordinates:

        "    -10,-8,20,  !- X,Y,Z ==> Vertex 1 {m}",
        "    -10,-8,0,  !- X,Y,Z ==> Vertex 2 {m}",
        "    10,-8,0,  !- X,Y,Z ==> Vertex 3 {m}",
        "    10,-8,20;  !- X,Y,Z ==> Vertex 4 {m}",

with the outer shade having coordinates:

        "    -10,-9,20,  !- X,Y,Z ==> Vertex 1 {m}",
        "    -10,-9,0,  !- X,Y,Z ==> Vertex 2 {m}",
        "    10,-9,0,  !- X,Y,Z ==> Vertex 3 {m}",
        "    10,-9,20;  !- X,Y,Z ==> Vertex 4 {m}",

So the outer shade is the same size as the inner shade which are larger than the wall. OK fine. These surfaces face North and location is Chicago, so the sun is behind these surfaces? And then seeing the SurfSunlitFrac = 1 for the mirrored shading surfaces would make sense? I still can't figure out why the wall and window SurfSunlitFrac = 0 when the shade transmittance = 1.

rraustad commented:
@Myoldmopar I have done as much as I can here.

jmarrec committed:
clang format

jmarrec commented:
@rraustad thank you very much!

jmarrec commented:
The SUNCOSTS being (0.2, 0.2, 0.2) is unrealistic.

Gotten from running 1ZoneUncontrolled.idf with chicago weather on Jan 1 at 12:

```
    state->dataBSDFWindow->SUNCOSTS(state->dataGlobal->TimeStep, state->dataGlobal->HourOfDay)(1) = 0.20531446332266728;
    state->dataBSDFWindow->SUNCOSTS(state->dataGlobal->TimeStep, state->dataGlobal->HourOfDay)(2) = -0.84761109808931534;
    state->dataBSDFWindow->SUNCOSTS(state->dataGlobal->TimeStep, state->dataGlobal->HourOfDay)(3) = 0.48928662105799514;
```

THe South Wall has a OutNormVec being OutNormVec = (x = 0, y = -1, z = 0), so when using 0.2 for the SUNCOSTS y, you end up with a negative number.

jmarrec committed:
Use relatistic SUNCOSTS

Gotten from running 1ZoneUncontrolled.idf with chicago weather on Jan 1 at 12

jmarrec commented:
Alright, I think everything should be green now.

There will be a conflict with #10657 in SolarShading.cc but that's easy to resolve (just use the #10657 version of CLIPLINE)

rraustad commented:
Thanks @jmarrec that did the trick. The SurfSunlitFrac values are back to the original values. However, I am confused as to how these unit tests passed when the SUNCOS data was not provided.

jmarrec commented:
@rraustad We shouldn't be too concerned about this (we shouldn't play wiht NaN to begin with), but to satisfy the curiosity:

In https://github.com/NREL/EnergyPlus/blob/52fee4b00a1d679d545dfaf69a3704227b11ff4c/src/EnergyPlus/SolarShading.cc#L5022-L5032

IEEE 754 says that operator < returns false 

> As specified, the predicates associated with the <, ≤, =, ≥, > mathematical symbols (or equivalent notation in programming languages) return false on an unordered relation

https://en.wikipedia.org/wiki/NaN

A functionally equivalent test code is this: https://godbolt.org/z/sq5nxj6Gz


state.dataSolarShading->SurfSunCosTheta is NaN as well.



cf https://github.com/NREL/EnergyPlus/blob/52fee4b00a1d679d545dfaf69a3704227b11ff4c/third_party/ObjexxFCL/src/ObjexxFCL/TypeTraits.hh#L633-L666

rraustad reviewed (approved):
I look forward to using this in the future.

Myoldmopar committed:
Merged develop; resolved conflicts

Myoldmopar commented:
Alright, I pulled in develop, choosing the other PR's changes in SolarShading.  Everything passes nicely, but I will let CI confirm this once more.  Thanks @jmarrec and @rraustad !

Myoldmopar commented:
This is really exciting, thanks for all the effort @rraustad and @jmarrec !  Tested it locally with develop one last time, and it's good.  Super happy on CI. Merging.
